metricbeat/helper/prometheus/textparse.go

```diff
@@ -0,0 +1,694 @@
+
+package prometheus
+
+import (
+	"math"
+	"mime"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/exemplar"
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/pkg/timestamp"
+)
+
+type Gauge struct {
+	Value *float64
+}
+
+func (m *Gauge) GetValue() float64 {
+	if m != nil && m.Value != nil {
+		return *m.Value
+	}
+	return 0
+}
+
+type Info struct {
+	Value *int64
+}
+
+func (m *Info) GetValue() int64 {
+	if m != nil && m.Value != nil {
+		return *m.Value
+	}
+	return 0
+}
+func (m *Info) HasValidValue() bool {
+	return m != nil && *m.Value == 1
+}
+
+type Stateset struct {
+	Value *int64
+}
+
+func (m *Stateset) GetValue() int64 {
+	if m != nil && m.Value != nil {
+		return *m.Value
+	}
+	return 0
+}
+func (m *Stateset) HasValidValue() bool {
+	return m != nil && (*m.Value == 0 || *m.Value == 1)
+}
+
+type Counter struct {
+	Value *float64
+}
+
+func (m *Counter) GetValue() float64 {
+	if m != nil && m.Value != nil {
+		return *m.Value
+	}
+	return 0
+}
+
+type Quantile struct {
+	Quantile *float64
+	Value    *float64
+	Exemplar *exemplar.Exemplar
+}
+
+func (m *Quantile) GetQuantile() float64 {
+	if m != nil && m.Quantile != nil {
+		return *m.Quantile
+	}
+	return 0
+}
+
+func (m *Quantile) GetValue() float64 {
+	if m != nil && m.Value != nil {
+		return *m.Value
+	}
+	return 0
+}
+
+type Summary struct {
+	SampleCount *uint64
+	SampleSum   *float64
+	Quantile    []*Quantile
+}
+
+func (m *Summary) GetSampleCount() uint64 {
+	if m != nil && m.SampleCount != nil {
+		return *m.SampleCount
+	}
+	return 0
+}
+
+func (m *Summary) GetSampleSum() float64 {
+	if m != nil && m.SampleSum != nil {
+		return *m.SampleSum
+	}
+	return 0
+}
+
+func (m *Summary) GetQuantile() []*Quantile {
+	if m != nil {
+		return m.Quantile
+	}
+	return nil
+}
+
+type Unknown struct {
+	Value *float64
+}
+
+func (m *Unknown) GetValue() float64 {
+	if m != nil && m.Value != nil {
+		return *m.Value
+	}
+	return 0
+}
+
+type Bucket struct {
+	CumulativeCount *uint64
+	UpperBound      *float64
+	Exemplar        *exemplar.Exemplar
+}
+
+func (m *Bucket) GetCumulativeCount() uint64 {
+	if m != nil && m.CumulativeCount != nil {
+		return *m.CumulativeCount
+	}
+	return 0
+}
+
+func (m *Bucket) GetUpperBound() float64 {
+	if m != nil && m.UpperBound != nil {
+		return *m.UpperBound
+	}
+	return 0
+}
+
+type Histogram struct {
+	SampleCount      *uint64
+	SampleSum        *float64
+	Bucket           []*Bucket
+	IsGaugeHistogram bool
+}
+
+func (m *Histogram) GetSampleCount() uint64 {
+	if m != nil && m.SampleCount != nil {
+		return *m.SampleCount
+	}
+	return 0
+}
+
+func (m *Histogram) GetSampleSum() float64 {
+	if m != nil && m.SampleSum != nil {
+		return *m.SampleSum
+	}
+	return 0
+}
+
+func (m *Histogram) GetBucket() []*Bucket {
+	if m != nil {
+		return m.Bucket
+	}
+	return nil
+}
+
+type OpenMetric struct {
```

Comment:
Nit. Could `OpenMetric` be an interface instead, and have individual types for each type of metric?

Reply:
Sure! I think it's the same with https://github.com/elastic/beats/pull/33865#discussion_r1050466560